### PR TITLE
Added new server config flag `--bootnode-only-sync` to ignore non-bootnodes in BestPeer()

### PR DESCRIPTION
**Pull request by dankostiuk**
# Description

When testing potential breaking changes on our testnets, we occasionally must stop all nodes and revert to a recent backup in order to revert the blockchain state to what it was prior to the breaking change. This process is fairly straightforward while dealing with a chain entirely operated by us. 

The issue lies when other peers or validators are connected to our mainnet chain which we do not have control over. As a precaution, if a new version of edge containing a breaking change was introduced on mainnet, it would be difficult to perform any kind of revert as our bootnodes would always sync to a peer that could have a higher blockNumber than the backup we want to revert to.

This PR introduces a new flag `--bootnode-only-sync` that we could require all validators to add after clearing their data-dirs to facilitate the unlikely event of having to reset to an earlier state. Additionally we would also add this flag to our bootnodes.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [ ] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

1. Spin up 4-node testnet with all 4 nodes as bootnodes
2. Add a 5th node to the testnet and give enough time to sync up
3. Clear the data (`rm -rf data/blockchain/ data/trie/ data/consensus/snapshots data/consensus/metadata`) of node 1.
4. Restart node 1 with `./main server --bootnode-only-sync` and observe that the non-bootnode peer is ignored when determining best peer within BestPeer()

# Documentation update

Will create a documentation PR once this PR is accepted.

### CODE REVIEW

**Comment by mrwillis** [comment]
The PR just blocks syncing from any peer but the bootnodes.

**Comment by Kourin1996** [comment]
@dankostiuk Hello, Thank you for a PR! We've refactored syncer recently. Could you fetch and merge latest `develop` if you've not done? Sorry for the inconvenience.

File: network/server.go [hunk 1]
```diff
@@ -523,6 +523,17 @@ func (s *Server) updateBootnodeConnCount(peerID peer.ID, delta int64) {
...
+// ShouldIgnoreSyncToPeer Check if client should ignore this peer when determining best peer to sync to
+func (s *Server) ShouldIgnoreSyncToPeer(peerID peer.ID) bool {
+	// If the peer is not a bootnode and the client has BootnodeOnlySync enabled
+	ignoreSync := s.config.BootnodeOnlySync && !s.bootnodes.isBootnode(peerID)
+	if ignoreSync {
+		s.logger.Debug("avoid syncing to non-bootnode peer", "id", peerID.String())
+	}
+
+	return ignoreSync
+}
```
Comment (Kourin1996) [inline network/server.go:535]: Right now only syncer syncs some data between peers. I think network layer shouldn't have such a function. Network layer just provides a method to return a peer is bootnode or not.

File: syncer/syncer.go [hunk 2]
```diff
@@ -209,10 +211,24 @@ func (s *syncer) BulkSync(newBlockCallback func(*types.Block) bool) error {
 	return nil
 }
 
+// makeSkipList initializes list of peers to skip
+func (s *syncer) makeSkipList() map[peer.ID]bool {
+	skipList := make(map[peer.ID]bool)
+
+	s.peerMap.Range(func(key, value interface{}) bool {
+		peer, _ := value.(*NoForkPeer)
+		skipList[peer.ID] = s.network.ShouldIgnoreSyncToPeer(peer.ID)
```
Comment (Kourin1996) [inline syncer/syncer.go:220]: Is it possible to rewrite this code by calling a method to given peer is bootnode or not?

**Review by Kourin1996 — CHANGES_REQUESTED** [review]

**Review by lazartravica — CHANGES_REQUESTED** [review]
Hello Dan,

I'm very sorry but I would need to close this Pull Request, as this would introduce a new class of "privileged" nodes (in this case bootnodes) that can be allowed to send blocks.

I understand the issue with coordinating breaking changes and upgrades, but this needs to be solved on the organizational side, not on the side of changing the syncing logic.

**Pull request closed without merging**